### ISSUE TITLE
running pytest with cuda

### ISSUE THREAD

**Issue by haruishi43** [open-mmlab/mmcv#1070]
**Describe the Issue**

I'm playing around with `mmcv` and was writing some debugging code for `pytest` when I came across `AssertionError` for `test_tin_shift.py` (same as [628](https://github.com/open-mmlab/mmcv/issues/628#issuecomment-723917559)).
I was running the full test at that time (`pytest tests`), but whenever I run the specific test (`pytest tests/test_ops/test_tin_shift.py`), it passes.

**Reproduction**

1. What command, code, or script did you run?

```bash
$ pytest tests  # fails
$ pytest tests/test_ops/test_tin_shift.py  # passes
```

2. Did you make any modifications on the code? Did you understand what you have modified?

- Modified `setup.py` so that it can compile on ampere architectures

**Environment**

```
{'sys.platform': 'linux', 'Python': '3.8.6 (default, Mar 15 2021, 08:31:51) [GCC 7.5.0]', 'CUDA available': True, 'GPU 0': 'GeForce RTX 3090', 'CUDA_HOME': '/usr/local/cuda', 'NVCC': 'Build cuda_11.0_bu.TC445_37.28845127_0', 'GCC': 'gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0', 'PyTorch': '1.7.1+cu110', 'PyTorch compiling details': 'PyTorch built with:\n  - GCC 7.3\n  - C++ Version: 201402\n  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications\n  - Intel(R) MKL-DNN v1.6.0 (Git Hash 5ef631a030a6f73131c77892041042805a06064f)\n  - OpenMP 201511 (a.k.a. OpenMP 4.5)\n  - NNPACK is enabled\n  - CPU capability usage: AVX2\n  - CUDA Runtime 11.0\n  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80\n  - CuDNN 8.0.5\n  - Magma 2.5.2\n  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, \n', 'TorchVision': '0.8.2+cu110', 'OpenCV': '4.5.1', 'MMCV': '1.3.5', 'MMCV Compiler': 'GCC 7.5', 'MMCV CUDA Compiler': '11.0'}
```

**Error traceback**

```
_____________________________________________________________ test_tinshift[dtype0] _____________________________________________________________

dtype = torch.float32

    @pytest.mark.skipif(
        not torch.cuda.is_available(), reason='requires CUDA support')
    @pytest.mark.parametrize('dtype', [torch.float, torch.double, torch.half])
    def test_tinshift(dtype):
>       _test_tinshift_allclose(dtype=dtype)

tests/test_ops/test_tin_shift.py:105: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dtype = torch.float32

    def _test_tinshift_allclose(dtype):
        try:
            from mmcv.ops import tin_shift
        except ModuleNotFoundError:
            pytest.skip('TinShift op is not successfully compiled')
    
        for shift, output, grad in zip(shifts, outputs, grads):
            np_input = np.array(inputs)
            np_shift = np.array(shift)
            np_output = np.array(output)
            np_grad = np.array(grad)
    
            x = torch.tensor(
                np_input, dtype=dtype, device='cuda', requires_grad=True)
            shift = torch.tensor(np_shift, device='cuda').int()
    
            output = tin_shift(x, shift)
            output.backward(torch.ones_like(output))
>           assert np.allclose(
                output.data.type(torch.float).cpu().numpy(), np_output, 1e-3)
E           AssertionError: assert False
E            +  where False = <function allclose at 0x7f457ca0ff70>(array([[[[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n... ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]]]], dtype=float32), array([[[[ 0.4369, -3.7571],\n         [-1.1835, -1.6374],\n         [ 0.9534, -0.1321]],\n\n        [[-0.4658,  0.2162],\n...        [ 0.    ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]]]]), 0.001)
E            +    where <function allclose at 0x7f457ca0ff70> = np.allclose
E            +    and   array([[[[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n... ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]]]], dtype=float32) = <built-in method numpy of Tensor object at 0x7f443b86e4c0>()
E            +      where <built-in method numpy of Tensor object at 0x7f443b86e4c0> = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000...     [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]]).numpy
E            +        where tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000...     [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]]) = <built-in method cpu of Tensor object at 0x7f443b86e640>()
E            +          where <built-in method cpu of Tensor object at 0x7f443b86e640> = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000....0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0').cpu
E            +            where tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000....0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0') = <built-in method type of Tensor object at 0x7f443b86e640>(torch.float32)
E            +              where <built-in method type of Tensor object at 0x7f443b86e640> = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000....0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0').type
E            +                where tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000....0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0') = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000...       [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0',\n       grad_fn=<TINShiftFunctionBackward>).data
E            +              and   torch.float32 = torch.float

tests/test_ops/test_tin_shift.py:95: AssertionError
_____________________________________________________________ test_tinshift[dtype1] _____________________________________________________________

dtype = torch.float64

    @pytest.mark.skipif(
        not torch.cuda.is_available(), reason='requires CUDA support')
    @pytest.mark.parametrize('dtype', [torch.float, torch.double, torch.half])
    def test_tinshift(dtype):
>       _test_tinshift_allclose(dtype=dtype)

tests/test_ops/test_tin_shift.py:105: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dtype = torch.float64

    def _test_tinshift_allclose(dtype):
        try:
            from mmcv.ops import tin_shift
        except ModuleNotFoundError:
            pytest.skip('TinShift op is not successfully compiled')
    
        for shift, output, grad in zip(shifts, outputs, grads):
            np_input = np.array(inputs)
            np_shift = np.array(shift)
            np_output = np.array(output)
            np_grad = np.array(grad)
    
            x = torch.tensor(
                np_input, dtype=dtype, device='cuda', requires_grad=True)
            shift = torch.tensor(np_shift, device='cuda').int()
    
            output = tin_shift(x, shift)
            output.backward(torch.ones_like(output))
>           assert np.allclose(
                output.data.type(torch.float).cpu().numpy(), np_output, 1e-3)
E           AssertionError: assert False
E            +  where False = <function allclose at 0x7f457ca0ff70>(array([[[[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n... ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]]]], dtype=float32), array([[[[ 0.4369, -3.7571],\n         [-1.1835, -1.6374],\n         [ 0.9534, -0.1321]],\n\n        [[-0.4658,  0.2162],\n...        [ 0.    ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]]]]), 0.001)
E            +    where <function allclose at 0x7f457ca0ff70> = np.allclose
E            +    and   array([[[[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n... ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]]]], dtype=float32) = <built-in method numpy of Tensor object at 0x7f443b85fc80>()
E            +      where <built-in method numpy of Tensor object at 0x7f443b85fc80> = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000...     [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]]).numpy
E            +        where tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000...     [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]]) = <built-in method cpu of Tensor object at 0x7f443b85f740>()
E            +          where <built-in method cpu of Tensor object at 0x7f443b85f740> = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000....0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0').cpu
E            +            where tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000....0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0') = <built-in method type of Tensor object at 0x7f443b96dac0>(torch.float32)
E            +              where <built-in method type of Tensor object at 0x7f443b96dac0> = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000... 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0', dtype=torch.float64).type
E            +                where tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000... 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0', dtype=torch.float64) = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000...000],\n          [ 0.0000,  0.0000]]]], device='cuda:0', dtype=torch.float64,\n       grad_fn=<TINShiftFunctionBackward>).data
E            +              and   torch.float32 = torch.float

tests/test_ops/test_tin_shift.py:95: AssertionError
_____________________________________________________________ test_tinshift[dtype2] _____________________________________________________________

dtype = torch.float16

    @pytest.mark.skipif(
        not torch.cuda.is_available(), reason='requires CUDA support')
    @pytest.mark.parametrize('dtype', [torch.float, torch.double, torch.half])
    def test_tinshift(dtype):
>       _test_tinshift_allclose(dtype=dtype)

tests/test_ops/test_tin_shift.py:105: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dtype = torch.float16

    def _test_tinshift_allclose(dtype):
        try:
            from mmcv.ops import tin_shift
        except ModuleNotFoundError:
            pytest.skip('TinShift op is not successfully compiled')
    
        for shift, output, grad in zip(shifts, outputs, grads):
            np_input = np.array(inputs)
            np_shift = np.array(shift)
            np_output = np.array(output)
            np_grad = np.array(grad)
    
            x = torch.tensor(
                np_input, dtype=dtype, device='cuda', requires_grad=True)
            shift = torch.tensor(np_shift, device='cuda').int()
    
            output = tin_shift(x, shift)
            output.backward(torch.ones_like(output))
>           assert np.allclose(
                output.data.type(torch.float).cpu().numpy(), np_output, 1e-3)
E           AssertionError: assert False
E            +  where False = <function allclose at 0x7f457ca0ff70>(array([[[[ 0.        ,  0.        ],\n         [ 0.        ,  0.        ],\n         [ 0.        ,  0.        ]],\n\n     ...[ 0.        ,  0.        ],\n         [ 0.        ,  0.        ],\n         [ 0.        ,  0.        ]]]], dtype=float32), array([[[[ 0.4369, -3.7571],\n         [-1.1835, -1.6374],\n         [ 0.9534, -0.1321]],\n\n        [[-0.4658,  0.2162],\n...        [ 0.    ,  0.    ]],\n\n        [[ 0.    ,  0.    ],\n         [ 0.    ,  0.    ],\n         [ 0.    ,  0.    ]]]]), 0.001)
E            +    where <function allclose at 0x7f457ca0ff70> = np.allclose
E            +    and   array([[[[ 0.        ,  0.        ],\n         [ 0.        ,  0.        ],\n         [ 0.        ,  0.        ]],\n\n     ...[ 0.        ,  0.        ],\n         [ 0.        ,  0.        ],\n         [ 0.        ,  0.        ]]]], dtype=float32) = <built-in method numpy of Tensor object at 0x7f443b972ec0>()
E            +      where <built-in method numpy of Tensor object at 0x7f443b972ec0> = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000...     [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]]).numpy
E            +        where tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000...     [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]]) = <built-in method cpu of Tensor object at 0x7f443b972cc0>()
E            +          where <built-in method cpu of Tensor object at 0x7f443b972cc0> = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000....0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0').cpu
E            +            where tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000....0000]],\n\n         [[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0') = <built-in method type of Tensor object at 0x7f443b96d5c0>(torch.float32)
E            +              where <built-in method type of Tensor object at 0x7f443b96d5c0> = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000... 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0', dtype=torch.float16).type
E            +                where tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000... 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]]]], device='cuda:0', dtype=torch.float16) = tensor([[[[ 0.0000,  0.0000],\n          [ 0.0000,  0.0000],\n          [ 0.0000,  0.0000]],\n\n         [[ 0.0000,  0.000...000],\n          [ 0.0000,  0.0000]]]], device='cuda:0', dtype=torch.float16,\n       grad_fn=<TINShiftFunctionBackward>).data
E            +              and   torch.float32 = torch.float

tests/test_ops/test_tin_shift.py:95: AssertionError
```

**Bug fix**

- Somehow while running the full test I was watching GPU memory usage using `nvidia-smi`, the memory usage seemed a lot (I don't know how much memory consumption it should be for each tests, but for `test_ms_deformable_attn.py`, the `2048` and `3096` channel tests didn't pass due to OOM). I thought it might have to do with `torch` not clearing cache during the full `pytest` run, so I added `torch.cuda.empty_cache()` to each of the tests in `test_ops` to see if the memory usage becomes lower, but nothing changed. It's so weird that the single test passes, but running all the tests fails.

**Comment by ZwwWayne** [comment]
Please @dreamerlin have a look, who wrote this part.

**Comment by zhouzaida** [comment]
Fixed by #1426
